# HG changeset patch
# User John W. Eaton <jwe@octave.org>
# Date 1284633015 14400
#      Thu Sep 16 06:30:15 2010 -0400
# Node ID 20bb05b9fb3f9648c6cb643ddc324999794efac9
# Parent  165e7e79b82c8eb0e20b1b92b9a27d2f625618b6
makefile rules for compiled auxiliary programs

diff --git a/ChangeLog b/ChangeLog
--- a/ChangeLog
+++ b/ChangeLog
@@ -1,8 +1,30 @@
+2010-09-16  John W. Eaton  <jwe@octave.org>
+
+	* configure.ac (SCRIPTS_EXE_SUFFIX) Delete variable and all uses.
+	(BUILD_COMPILED_AUX_PROGRAMS): New variable.
+	(AMCOND_BUILD_COMPILED_AUX_PROGRAMS): New Automake conditional.
+	* Makefile.am (SCRIPTS_EXE_SUFFIX, BUILT_SHELL_FILES): Delete
+	variables and all uses.
+	(bin_SCRIPTS, bin_PROGRAMS, mkoctfile_SOURCES,
+	nodist_mkoctfile_SOURCES, octave_config_SOURCES,
+	nodist_octave_config_SOURCES, octave_bug_SOURCES,
+	nodist_octave_bug_SOURCES, BUILT_SOURCES): Define conditionally
+	based on AMCOND_BUILD_COMPILED_AUX_PROGRAMS.
+	(noinst_SCRIPTS): New variable.
+	(CLEANFILES): Include $(bin_PROGRAMS) $(bin_SCRIPTS)
+	$(BUILT_SOURCES) in the list.
+	(all-local): Also depend on $(bin_PROGRAMS) and $(noinst_SCRIPTS).
+	(octave-bug.cc, octave-config.cc, mkoctfile.cc, octave-bug)
+	(octave-config, mkoctfile): Make rules conditional on
+	AMCOND_BUILD_COMPILED_AUX_PROGRAMS.
+	(make-version-links, remove-version-links): Define different
+	rules conditional on AMCOND_BUILD_COMPILED_AUX_PROGRAMS.
+
 2010-09-13  Jaroslav Hajek  <highegg@gmail.com>
 
 	* NEWS: Update.
 
 2010-09-10  Jaroslav Hajek  <highegg@gmail.com>
 
 	* m4/ax_blas_f77_func.m4: Update.
 
diff --git a/Makefile.am b/Makefile.am
--- a/Makefile.am
+++ b/Makefile.am
@@ -22,18 +22,16 @@
 include common.mk
 
 ## Avoid making multiple subdirs in parallel which can lead 
 ## to a confusing error message stream
 .NOTPARALLEL:
 
 ACLOCAL_AMFLAGS = -I m4
 
-SCRIPTS_EXE_SUFFIX = @SCRIPTS_EXE_SUFFIX@
-
 BUILT_DISTFILES = BUGS INSTALL.OCTAVE
 
 EXTRA_DIST = \
   BUGS \
   COPYING \
   ChangeLog \
   ChangeLog.1 \
   INSTALL \
@@ -72,88 +70,89 @@ include m4/module.mk
 SUBDIRS = libgnu libcruft liboctave src scripts @DOCDIR@ examples test
 
 if AMCOND_BUILD_DOCS
 else
 dist-hook:
 	echo "Documentation disabled.  Cannot package distribution!" ; exit 1;
 endif
 
-BUILT_SHELL_FILES = \
-  mkoctfile$(SCRIPTS_EXE_SUFFIX) \
-  octave-config$(SCRIPTS_EXE_SUFFIX) \
-  octave-bug$(SCRIPTS_EXE_SUFFIX) \
+if AMCOND_BUILD_COMPILED_AUX_PROGRAMS
+bin_PROGRAMS = \
+  mkoctfile \
+  octave-config \
+  octave-bug
+
+mkoctfile_SOURCES =
+nodist_mkoctfile_SOURCES = mkoctfile.cc
+
+octave_config_SOURCES =
+nodist_octave_config_SOURCES = octave-config.cc
+
+octave_bug_SOURCES =
+nodist_octave_bug_SOURCES = octave-bug.cc
+
+BUILT_SOURCES = \
+  mkoctfile.cc \
+  octave-config.cc \
+  octave-bug.cc
+
+else
+bin_SCRIPTS = \
+  mkoctfile \
+  octave-config \
+  octave-bug \
   run-octave
+endif
 
-bin_SCRIPTS = \
-  mkoctfile$(SCRIPTS_EXE_SUFFIX) \
-  octave-config$(SCRIPTS_EXE_SUFFIX) \
-  octave-bug$(SCRIPTS_EXE_SUFFIX)
+noinst_SCRIPTS = run-octave
 
 INFO_FILES = \
   INSTALL.OCTAVE \
   BUGS
 
-CLEANFILES = $(BUILT_SHELL_FILES)
+CLEANFILES = $(bin_PROGRAMS) $(bin_SCRIPTS) $(BUILT_SOURCES)
 
 DISTCLEANFILES = $(INFO_FILES) .gdbinit 
 
 CONFIG_FILES = @ac_config_headers@ @ac_config_files@
 
-BUILT_SOURCES = run-octave
-
 nodist_octinclude_HEADERS = config.h
 
-all-local: $(bin_SCRIPTS) $(INFO_FILES) .gdbinit
+all-local: $(bin_PROGRAMS) $(bin_SCRIPTS) $(noinst_SCRIPTS) $(INFO_FILES) .gdbinit
 	@echo ""
 	@echo "Octave successfully built.  Now choose from the following:"
 	@echo ""
 	@echo "   ./run-octave    - to run in place to test before installing"
 	@echo "   make check      - to run the tests"
 	@echo "   make install    - to install (PREFIX=$(prefix))"
 	@echo ""
 
+if AMCOND_BUILD_COMPILED_AUX_PROGRAMS
+octave-bug.cc: octave-bug.cc.in Makefile
+	@$(do_subst_config_vals)
+
+octave-config.cc: octave-config.cc.in Makefile
+	@$(do_subst_default_vals)
+
+mkoctfile.cc: mkoctfile.cc.in Makefile
+	@$(do_subst_config_vals)
+else
 octave-bug: octave-bug.in Makefile
 	@$(do_subst_config_vals)
 	chmod a+rx $@
 
-octave-bug.cc: octave-bug.cc.in Makefile
-	@$(do_subst_config_vals)
-
-# ifneq ($(EXEEXT),)
-# octave-bug$(EXEEXT): octave-bug.o
-# 	$(LD_CXX) $(CPPFLAGS) $(ALL_CXXFLAGS) $(RDYNAMIC_FLAG) \
-# 	    $(ALL_LDFLAGS) -o $@ octave-bug.o
-# endif
-
 octave-config: octave-config.in Makefile
 	@$(do_subst_default_vals)
 	chmod a+rx $@
 
-octave-config.cc: octave-config.cc.in Makefile
-	@$(do_subst_default_vals)
-
-# ifneq ($(EXEEXT),)
-# octave-config$(EXEEXT): octave-config.o
-# 	$(LD_CXX) $(CPPFLAGS) $(ALL_CXXFLAGS) $(RDYNAMIC_FLAG) \
-# 	    $(ALL_LDFLAGS) -o $@ octave-config.o
-# endif
-
 mkoctfile: mkoctfile.in Makefile
 	@$(do_subst_config_vals)
 	chmod a+rx $@
-
-mkoctfile.cc: mkoctfile.cc.in Makefile
-	@$(do_subst_config_vals)
-
-# ifneq ($(EXEEXT),)
-# mkoctfile$(EXEEXT): mkoctfile.o
-# 	$(LD_CXX) $(CPPFLAGS) $(ALL_CXXFLAGS) $(RDYNAMIC_FLAG) \
-# 	    $(ALL_LDFLAGS) -o $@ mkoctfile.o
-# endif
+endif
 
 run-octave: run-octave.in Makefile
 	@$(do_subst_script_vals)
 	chmod a+rx "$@"
 
 .gdbinit: gdbinit
 	if [ -f .gdbinit ]; then \
 	  echo "refusing to overwrite .gdbinit with newer version from $<" 1>&2; \
@@ -182,22 +181,36 @@ installdirs-local:
 	$(MKDIR_P) $(addprefix $(DESTDIR), $(DIRS_TO_MAKE))
 
 install-data-local: installdirs-local
 
 install-exec-hook: make-version-links
 
 uninstall-local: remove-version-links
 
+if AMCOND_BUILD_COMPILED_AUX_PROGRAMS
+make-version-links:
+	cd $(DESTDIR)$(bindir) && \
+	for f in $(basename $(bin_PROGRAMS)); do \
+	  mv $$f$(EXEEXT) $$f-$(version)$(EXEEXT) && \
+	    $(LN_S) $$f-$(version)$(EXEEXT) $$f$(EXEEXT); \
+	done
+
+remove-version-links:
+	for f in $(basename $(bin_PROGRAMS)); do \
+	  rm -f $(DESTDIR)$(bindir)/$$f-$(version)$(EXEEXT); \
+	done
+else
 make-version-links:
 	cd $(DESTDIR)$(bindir) && \
 	for f in $(basename $(bin_SCRIPTS)); do \
-	  mv $$f$(SCRIPTS_EXE_SUFFIX) $$f-$(version)$(SCRIPTS_EXE_SUFFIX) && \
-	    $(LN_S) $$f-$(version)$(SCRIPTS_EXE_SUFFIX) $$f$(SCRIPTS_EXE_SUFFIX); \
+	  mv $$f $$f-$(version) && \
+	    $(LN_S) $$f-$(version) $$f; \
 	done
-.PHONY: make-version-links
 
 remove-version-links:
 	for f in $(basename $(bin_SCRIPTS)); do \
-	  rm -f $(DESTDIR)$(bindir)/$$f-$(version)$(SCRIPTS_EXE_SUFFIX); \
+	  rm -f $(DESTDIR)$(bindir)/$$f-$(version); \
 	done
-.PHONY: remove-version-links
+endif
 
+.PHONY: make-version-links remove-version-links
+
diff --git a/configure.ac b/configure.ac
--- a/configure.ac
+++ b/configure.ac
@@ -1146,17 +1146,17 @@ NO_OCT_FILE_STRIP=false
 TEMPLATE_AR='$(AR)'
 TEMPLATE_ARFLAGS="$ARFLAGS"
 CRUFT_DLL_DEFS=
 OCTAVE_DLL_DEFS=
 OCTINTERP_DLL_DEFS=
 OCTGRAPHICS_DLL_DEFS=
 library_path_var=LD_LIBRARY_PATH
 ldpreloadsep=" "
-SCRIPTS_EXE_SUFFIX=
+BUILD_COMPILED_AUX_PROGRAMS=false
 case "$canonical_host_type" in
   *-*-386bsd* | *-*-netbsd*)
     SH_LD=ld
     SH_LDFLAGS=-Bshareable
   ;;
   *-*-openbsd*)
     SH_LDFLAGS='-shared -fPIC'
   ;;
@@ -1201,16 +1201,17 @@ case "$canonical_host_type" in
     SHLLIB=dll.a
     SHLBIN=dll    
     DL_LDFLAGS="-shared -Wl,--export-all-symbols -Wl,--enable-auto-import -Wl,--enable-runtime-pseudo-reloc"
     SH_LDFLAGS="-shared -Wl,--export-all-symbols -Wl,--enable-auto-import -Wl,--enable-auto-image-base"
     SONAME_FLAGS='-Wl,--out-implib=$(patsubst $(SHLPRE)%,$(LIBPRE)%,$@).a'
     ldpreloadsep=":"
   ;;
   *-*-mingw*)
+    BUILD_COMPILED_AUX_PROGRAMS=true
     if test "$have_msvc" = "yes"; then
       DL_LDFLAGS="-shared"
       CPICFLAG=
       CXXPICFLAG=
       FPICFLAG=
       SHLEXT=dll
       SHLLIB=lib
       SHLBIN=dll
@@ -1226,33 +1227,32 @@ case "$canonical_host_type" in
       NO_OCT_FILE_STRIP=true
       library_path_var=PATH
       NO_OCT_FILE_STRIP=true
       ## Extra compilation flags.
       CRUFT_DLL_DEFS="-DCRUFT_DLL"
       OCTAVE_DLL_DEFS="-DOCTAVE_DLL"
       OCTINTERP_DLL_DEFS="-DOCTINTERP_DLL"
       OCTGRAPHICS_DLL_DEFS="-DOCTGRAPHICS_DLL"
-      SCRIPTS_EXE_SUFFIX='$(EXEEXT)'
     else
       CPICFLAG=
       CXXPICFLAG=
       FPICFLAG=
       SHLEXT=dll
       SHLLIB=dll.a
       SHLBIN=dll
       DL_LDFLAGS="-shared -Wl,--export-all-symbols -Wl,--enable-auto-import -Wl,--enable-runtime-pseudo-reloc"
       SH_LDFLAGS="-shared -Wl,--export-all-symbols -Wl,--enable-auto-import -Wl,--enable-auto-image-base"
       SONAME_FLAGS='-Wl,--out-implib=$@.a'
       library_path_var=PATH
-      SCRIPTS_EXE_SUFFIX='$(EXEEXT)'
     fi
   ;;
 
   *-*-msdosmsvc)
+    BUILD_COMPILED_AUX_PROGRAMS=true
     DL_LDFLAGS="-shared"
     CPICFLAG=
     CXXPICFLAG=
     FPICFLAG=
     SHLEXT=dll
     SHLLIB=lib
     SHLBIN=dll
     LIBPRE=
@@ -1267,17 +1267,16 @@ case "$canonical_host_type" in
     NO_OCT_FILE_STRIP=true
     library_path_var=PATH
     NO_OCT_FILE_STRIP=true
     ## Extra compilation flags.
     CRUFT_DLL_DEFS="-DCRUFT_DLL"
     OCTAVE_DLL_DEFS="-DOCTAVE_DLL"
     OCTINTERP_DLL_DEFS="-DOCTINTERP_DLL"
     OCTGRAPHICS_DLL_DEFS="-DOCTGRAPHICS_DLL"
-    SCRIPTS_EXE_SUFFIX='$(EXEEXT)'
   ;;
   *-*-linux* | *-*-gnu*)
     MKOCTFILE_DL_LDFLAGS="-shared -Wl,-Bsymbolic"
     SONAME_FLAGS='-Wl,-soname -Wl,$@'
     RLD_FLAG='-Wl,-rpath -Wl,$(octlibdir)'
   ;;
   i[[3456]]86-*-sco3.2v5*)
     SONAME_FLAGS='-Wl,-h -Wl,$@'
@@ -1350,16 +1349,19 @@ case "$canonical_host_type" in
       true
     else
       TEMPLATE_AR='$(CXX)'
       TEMPLATE_ARFLAGS="-xar -o"
     fi
   ;;
 esac
 
+AM_CONDITIONAL([AMCOND_BUILD_COMPILED_AUX_PROGRAMS],
+  [test x$BUILD_COMPILED_AUX_PROGRAMS = xtrue])
+
 if $use_rpath; then
   if test -n "$enable_rpath_arg"; then
     RLD_FLAG="$enable_rpath_arg"
   fi
 else
   RLD_FLAG=""
 fi
 
@@ -1417,17 +1419,16 @@ AC_SUBST(RLD_FLAG)
 AC_SUBST(TEMPLATE_AR)
 AC_SUBST(TEMPLATE_ARFLAGS)
 AC_SUBST(CRUFT_DLL_DEFS)
 AC_SUBST(OCTAVE_DLL_DEFS)
 AC_SUBST(OCTINTERP_DLL_DEFS)
 AC_SUBST(OCTGRAPHICS_DLL_DEFS)
 AC_SUBST(library_path_var)
 AC_SUBST(ldpreloadsep)
-AC_SUBST(SCRIPTS_EXE_SUFFIX)
 
 ### special checks for odd OS specific things.
 
 AC_CHECK_FUNCS(getpwnam, [], [AC_CHECK_LIB(sun, getpwnam)])
 
 NO_UNDEFINED_LDFLAG=
 case "$canonical_host_type" in
   *-*-cygwin*)
